modin/pandas/dataframe.py

```diff
@@ -87,6 +88,7 @@ def __init__(
         query_compiler: query_compiler
             A query compiler object to manage distributed computation.
         """
+        Engine.subscribe(_update_engine)
```

Comment:
By updating the engine at every call to dataframe, we have undefined (or at least undocumented) behavior when a user decides to change the engine on-the-fly. For example,
```
import modin.pandas as pd

# Build dataframe with ray engine.
os.environ["MODIN_ENGINE"] = "ray"
import ray
ray.init()

df = pd.DataFrame([[1,2,3],[4,5,6]])

# Change to dask engine.
os.environ["MODIN_ENGINE"] = "dask"

df.sum()
```

Right now, it runs the `df.sum()` and returns a result. I am assuming that it is running with the ray engine. Do we want this to be the intended behavior? If so, we should move this line to only be triggered when there is `query_compiler` is None (i.e., when we are creating the dataframe). The same should be done for series.

Reply:
Just noticed this, but doesn't all of the IO functions eventually hit this line as well?

Reply:
The execution of each dataframe is tied to it when it is created. All child dataframes inherit the execution context, so the behavior is defined but undocumented. I actually don't want users to do this, because you won't be able to join the dataframes.

Reply:
Which IO functions are not covered? They all should be.

Reply:
> The execution of each dataframe is tied to it when it is created. All child dataframes inherit the execution context, so the behavior is defined but undocumented. I actually don't want users to do this, because you won't be able to join the dataframes.
Ah ok that makes sense. I mistakenly believed that the `__init__` would be called for every function.

> Which IO functions are not covered? They all should be.
I believe they all are. I wanted to find a way to move the `Engine.subscribe(_update_engine)` to one place as there may be other things we want to do when initializing new dataframes. However, we can push that off until later when we know we have something else that needs to be done.

Reply:
I don't think changing the execution engine on the fly is a good idea. I would prefer to prescribe and document the engine set at the beginning of the session. I cannot think of a good example when you need to change the engine in the middle of the processing.

Reply:
The engine is subscribed to exactly once. It is the same as importing modin multiple times in the current iteration. Even if the flag is changed, the engine will not be updated for existing or new dataframes, it is only set the first time the user does any operation.

This example shows that changing the `MODIN_ENGINE` does not change anything while computing, only when the first engine is subscribed to does any update internally happen. After that there is no update

```python
import modin.pandas as pd
import os

os.environ["MODIN_ENGINE"] = "dask"
df = pd.DataFrame([1,2,3])
os.environ["MODIN_ENGINE"] = "ray"
df = pd.DataFrame([3,4,5])  # still dask
```

We document that you should not change the `"MODIN_ENGINE"` in the middle of computation, but it is still safe to do so in that we will just ignore it.

Reply:
> Which IO functions are not covered? They all should be.

Why `read_clipboard` and `read_feather` are not covered?

Reply:
@YarShev they should be. I am adding them